lib/spack/docs/packaging_guide.rst

```diff
@@ -480,6 +480,12 @@ like so:
 .. literalinclude:: _spack_root/var/spack/repos/builtin/packages/openmpi/package.py
    :pyobject: Openmpi.url_for_version
 
+.. note::
+
+   If only the extension of the source archive changes, it is easier to use `extension`
+   argument to `version()`
```

Comment:
I added `extension` to `fetch_strategy`. I didn't realize it could be used this way: my understanding was if you have a url like

`https://example.com/foo-1.1.tar.gz`

and the URL for version `1.2` is

`https://example.com/foo-1.2.zip`

That setting `extension` on the `version()` will account for that? I didn't expect that from looking at `fetch_strategy` again, but if I'm wrong about that do you have an example (either an existing Spack package or a PR where this works)?

Thanks!

Reply:
E.g. `giflib`:
```
$ spack fetch giflib@5.2.1
==> Fetching https://mirror.spack.io/_source-cache/archive/31/31da5562f44c5f15d63340a09a4fd62b48c45620cd302f77a6d9acf0077879bd.tar.gz
################################################################################################################# 100.0%
$ spack fetch giflib@5.1.4
==> Fetching https://mirror.spack.io/_source-cache/archive/df/df27ec3ff24671f80b29e6ab1c4971059c14ac3db95406884fc26574631ba8d5.tar.bz2
################################################################################################################# 100.0%
```
Or, without Spack mirror:
```
$ spack clean -d
$ spack fetch giflib@5.2.1
==> Fetching https://prdownloads.sourceforge.net/giflib/giflib-5.2.1.tar.gz
#################################################################################################################100.0%
################################################################################################################# 100.0%
$ spack fetch giflib@5.1.4
==> Fetching https://prdownloads.sourceforge.net/giflib/giflib-5.1.4.tar.bz2
#################################################################################################################100.0%
################################################################################################################# 100.0%
$ ls -R var/spack/cache/_source-cache/archive
var/spack/cache/_source-cache/archive:
31  df

var/spack/cache/_source-cache/archive/31:
31da5562f44c5f15d63340a09a4fd62b48c45620cd302f77a6d9acf0077879bd.tar.gz

var/spack/cache/_source-cache/archive/df:
df27ec3ff24671f80b29e6ab1c4971059c14ac3db95406884fc26574631ba8d5.tar.bz2
```
(however URLs shown by `spack info giflib` are not taking extension into account - but I guess this should be reported separately)

Reply:
The thing that makes me worried is that I have no idea how this is happening based on looking at the code (I also just looked through https://github.com/spack/spack/pull/1758 where `extension` was added and don't see how this works). There's a couple ways we could proceed to address that:

* write a unit test to enforce this behavior
* I could try to trace through an example of this (although I'm a bit short on time for that at the moment)
  * based on that, add more-deliberate support of this feature

I actually want to do both: even if this works for now, I'd worry that some minor seemingly-unrelated change could affect this behavior in the future.